File: neetcode150/06_linked_list/0025_reverse_nodes_in_k_group.py
"""
=====================================================================
LeetCode 25 : Reverse Nodes in k-Group                        (Hard)
https://leetcode.com/problems/reverse-nodes-in-k-group/
Category   : Linked List
---------------------------------------------------------------------
PROBLEM
    Reverse the nodes of a list k AT A TIME; the leftover tail
    (fewer than k nodes) stays as it is.
        1→2→3→4→5, k=2  →  2→1→4→3→5
        1→2→3→4→5, k=3  →  3→2→1→4→5
---------------------------------------------------------------------
INTUITION
    Cut the list into k-node blocks; each block is reversed with the
    exact prev/cur flip from 0206, then the blocks are sewn back
    together: block tail → next block head. A dummy node carries the
    pointer to the first group's new head.
---------------------------------------------------------------------
APPROACH 1 — Iterative, dummy + anchored segment flips (BEST)
    Peek k nodes ahead (fewer → stop, tail untouched); flip exactly
    k nodes after cursor; sew tail→next-group and cursor→new-head;
    cursor jumps to the block's old head (now its tail).
    Time  Complexity : O(n)   Space Complexity : O(1)
---------------------------------------------------------------------
APPROACH 2 — Recursive per group
    Reverse first k nodes; recurse on the remainder; sew this
    group's tail to the next group's reversed head.
    Time  Complexity : O(n)   Space Complexity : O(n/k) stack
---------------------------------------------------------------------
REAL LIFE
    • Block-wise transforms of page chains (pagination caches).
    • Batch reordering in distributed queues: reverse chunks of
      k records at a time, sew the stream back.
=====================================================================
"""

from __future__ import annotations
import time
import tracemalloc
from typing import Optional


class ListNode:
    """Minimal singly linked list node used by every test below."""

    def __init__(self, val: int = 0, next: Optional["ListNode"] = None) -> None:
        self.val = val
        self.next = next


def build_list(values: list[int]) -> Optional[ListNode]:
    """Turn a Python list into a linked list; returns the head."""
    dummy = ListNode()
    tail = dummy
    for v in values:
        tail.next = ListNode(v)
        tail = tail.next
    return dummy.next


def to_list(head: Optional[ListNode]) -> list[int]:
    """Walk a linked list back into a Python list (for verification)."""
    out: list[int] = []
    while head is not None:
        out.append(head.val)
        head = head.next
    return out


# =====================================================================
# APPROACH 1 : Iterative — dummy head + anchored segment flips (best)
# =====================================================================
class Solution_Iterative:
    """
    Purpose : Reverse the list in groups of exactly k nodes.
    Inputs  : head — first node of the list; k — group size ≥ 1.
    Output  : Head of the regrouped list (leftover tail unmodified).
    """

    def reverseKGroup(self, head: Optional[ListNode], k: int) -> Optional[ListNode]:
        dummy = ListNode(0, head)         # uniform sewing for group #1
        cursor: Optional[ListNode] = dummy   # node BEFORE current group

        while True:
            # Peek: does a full k-node group remain after cursor?
            probe = cursor
            for _ in range(k):
                probe = probe.next if probe else None
                if probe is None:
                    return dummy.next    # leftover tail stays as-is

            old_head = cursor.next        # becomes the group's TAIL
            prev: Optional[ListNode] = cursor   # flip target starts here
            cur = cursor.next             # first node to flip

            # Flip exactly k nodes (bounded 0206 flip).
            for _ in range(k):
                nxt = cur.next if cur else None   # CACHE successor
                if cur is not None:
                    cur.next = prev               # flip the arrow
                prev = cur
                cur = nxt

            # Sew: old_head (group tail) → next group; cursor → new head.
            if old_head is not None:
                old_head.next = cur
            if cursor is not None:
                cursor.next = prev
            cursor = old_head             # next group starts here


# =====================================================================
# APPROACH 2 : Recursive — reverse first group, recurse on the rest
# =====================================================================
class Solution_Recursive:
    """
    Purpose : Reverse the list in groups of exactly k nodes.
    Inputs  : head — first node of the list; k — group size ≥ 1.
    Output  : Head of the regrouped list.
    Note    : O(n/k) recursion depth.
    """

    def reverseKGroup(self, head: Optional[ListNode], k: int) -> Optional[ListNode]:
        # Probe: are there k nodes from here?
        probe = head
        for _ in range(k):
            if probe is None:
                return head              # fewer than k → keep as-is
            probe = probe.next

        # Reverse exactly k nodes (bounded flip).
        prev: Optional[ListNode] = None
        cur = head
        for _ in range(k):
            nxt = cur.next if cur else None
            if cur is not None:
                cur.next = prev
            prev = cur
            cur = nxt

        # head is now the group's TAIL → point it at the next
        # group's reversed head (recursion).
        if head is not None:
            head.next = self.reverseKGroup(cur, k)
        return prev                      # new head of this group


# =====================================================================
# BENCHMARK — 3 tests, both approaches, time + peak memory
# =====================================================================
if __name__ == "__main__":
    tests: list[tuple[list[int], int]] = [
        ([1, 2, 3, 4, 5], 2),
        ([1, 2, 3, 4, 5], 3),
        ([1, 2], 2),
    ]
    expected: list[list[int]] = [
        [2, 1, 4, 3, 5],
        [3, 2, 1, 4, 5],
        [2, 1],
    ]

    # --- Approach 1 ---
    tracemalloc.start()
    t0 = time.perf_counter()
    ok1 = all(
        to_list(Solution_Iterative().reverseKGroup(build_list(v), k)) == e
        for (v, k), e in zip(tests, expected)
    )
    t1 = time.perf_counter()
    _, peak1 = tracemalloc.get_traced_memory()
    tracemalloc.stop()

    # --- Approach 2 ---
    tracemalloc.start()
    t0b = time.perf_counter()
    ok2 = all(
        to_list(Solution_Recursive().reverseKGroup(build_list(v), k)) == e
        for (v, k), e in zip(tests, expected)
    )
    t1b = time.perf_counter()
    _, peak2 = tracemalloc.get_traced_memory()
    tracemalloc.stop()

    print("0025 Reverse Nodes in k-Group")
    print(f"Approach 1 (iterative) : time = {(t1 - t0) * 1e6:.2f} µs  peak-mem = {peak1} bytes")
    print(f"Approach 2 (recursive) : time = {(t1b - t0b) * 1e6:.2f} µs  peak-mem = {peak2} bytes")
    print("PASS : both approaches regrouped all 3 tests correctly."
          if ok1 and ok2 else "FAIL : at least one test produced a wrong list.")

File: neetcode150/06_linked_list/0025_reverse_nodes_in_k_group.cpp
/*
 * =====================================================================
 * LeetCode 25 : Reverse Nodes in k-Group                        (Hard)
 * https://leetcode.com/problems/reverse-nodes-in-k-group/
 * Category   : Linked List
 * ---------------------------------------------------------------------
 * PROBLEM
 *   Reverse the nodes of a list k AT A TIME. The leftover tail
 *   (fewer than k nodes) stays as it is. Reverse the direction of
 *   every group's arrows — the grouping itself never changes.
 *       1→2→3→4→5, k=2  →  2→1→4→3→5
 *       1→2→3→4→5, k=3  →  3→2→1→4→5
 * ---------------------------------------------------------------------
 * INTUITION
 *   Cut the list into k-node blocks. Each block is reversed with the
 *   EXACT prev/cur flip from 0206, but the block boundaries must be
 *   sewn back together: block_start->next = next_block_head, and the
 *   previous block's tail must point at this block's NEW head. A
 *   dummy node carries the pointer to the very first group's new head.
 * ---------------------------------------------------------------------
 * APPROACH 1 — Iterative with dummy + group cursor (BEST)
 *   dummy → head; cursor = dummy. Loop: peek k nodes ahead; if
 *   fewer than k remain, stop (leftover tail untouched). Otherwise
 *   reverse the k nodes right after cursor using the standard flip,
 *   then reattach: cursor->next->next = next_group (sew tail), and
 *   cursor->next = new_group_head. cursor jumps to the block's old
 *   head (now its tail).
 *   Time  Complexity : O(n)  — each node flipped exactly once.
 *   Space Complexity : O(1)  — pointers only.
 * ---------------------------------------------------------------------
 * APPROACH 2 — Recursive per group
 *   Reverse the first k nodes (they become the group's new head);
 *   recurse on the REMAINDER of the list, which returns the next
 *   group's reversed head; sew  tail_of_this_group->next = that.
 *   Base case: fewer than k nodes remain → return as-is.
 *   Time  Complexity : O(n)   Space Complexity : O(n/k) stack.
 * ---------------------------------------------------------------------
 * DEEP DIVE — reversing a BOUNDED segment between two fixed points
 *   Iterative k-group reversal is the "between markers" version of
 *   the flip: instead of flipping until null, flip exactly k nodes.
 *   The three anchors are:
 *       cursor  — node before the segment (never moves while
 *                 reversing inside);
 *       oldHead = cursor->next (first node of the segment, becomes
 *                 the segment's TAIL);
 *       prev/cur — the standard 0206 sliders doing the flips.
 *   After k flips, cur points at the next segment's first node, and
 *   oldHead's ->next must be set to it (sewing the chain back).
 *   This "reverse exactly K between anchors" pattern is the same
 *   one used by doubly-linked-page list manipulation and by
 *   block-wise transforms in databases. Get the anchor invariants
 *   right here and 0206/0143/0025 become one skill.
 * =====================================================================
 */

#include <vector>          // std::vector — test data
#include <chrono>          // std::chrono — timing
#include <iostream>        // std::cout — output
#include <sys/resource.h>  // getrusage — peak RAM
#include "list_node.h"     // our header: ListNode, buildList

/* =====================================================================
 * APPROACH 1 : Iterative — dummy head + anchored segment flips
 * ===================================================================== */
class Solution_Iterative {
public:
    /*
     * Purpose : Reverse the list in groups of exactly k nodes.
     * Inputs  : head — first node of the list; k — group size ≥ 1.
     * Output  : Head of the regrouped list (leftover tail unmodified).
     */
    ListNode* reverseKGroup(ListNode* head, int k) {
        // Dummy lets the FIRST group's new head be sewn in uniformly.
        ListNode* dummy = new ListNode(0, head);
        ListNode* cursor = dummy;   // node just BEFORE the current group

        while (true) {
            // Peek: does a full k-node group remain after cursor?
            ListNode* probe = cursor;
            for (int i = 0; i < k; ++i) {
                probe = probe->next;
                if (probe == nullptr) {
                    // Fewer than k nodes left → leftover tail stays.
                    ListNode* result = dummy->next;
                    delete dummy;
                    return result;
                }
            }

            // Segment anchors: oldHead is first of the group and will
            // end up as its TAIL after the reversal.
            ListNode* oldHead = cursor->next;
            ListNode* prev = cursor;      // flip target starts at cursor
            ListNode* cur = cursor->next; // node to flip first

            // Flip exactly k nodes (the 0206 flip, bounded).
            for (int i = 0; i < k; ++i) {
                ListNode* nxt = cur->next;  // CACHE successor
                cur->next = prev;           // flip the arrow
                prev = cur;                 // slide
                cur = nxt;
            }

            // Sewing: oldHead (now group tail) must point at the next
            // group's first node; cursor (before the group) points at
            // the group's NEW head (prev).
            oldHead->next = cur;
            cursor->next = prev;

            cursor = oldHead;  // next group starts right after oldHead
        }
    }
};

/* =====================================================================
 * APPROACH 2 : Recursive — reverse first group, recurse on the rest
 * ===================================================================== */
class Solution_Recursive {
public:
    /*
     * Purpose : Reverse the list in groups of exactly k nodes.
     * Inputs  : head — first node of the list; k — group size ≥ 1.
     * Output  : Head of the regrouped list.
     * Note    : O(n/k) recursion depth.
     */
    ListNode* reverseKGroup(ListNode* head, int k) {
        // Probe: are there k nodes from here?
        ListNode* probe = head;
        for (int i = 0; i < k; ++i) {
            if (probe == nullptr) {
                return head;   // fewer than k → this tail stays as-is
            }
            probe = probe->next;
        }

        // Reverse exactly k nodes (bounded 0206 flip).
        ListNode* prev = nullptr;
        ListNode* cur = head;
        for (int i = 0; i < k; ++i) {
            ListNode* nxt = cur->next;
            cur->next = prev;
            prev = cur;
            cur = nxt;
        }

        // head is now this group's TAIL → it must point at the
        // next group's reversed head (computed recursively).
        head->next = reverseKGroup(cur, k);
        return prev;   // new head of this group
    }
};

/* =====================================================================
 * BENCHMARK — 3 tests, both approaches, time + peak RAM
 * ===================================================================== */
int main() {
    std::vector<std::vector<int>> vals = {{1, 2, 3, 4, 5}, {1, 2, 3, 4, 5}, {1, 2}};
    std::vector<int> k = {2, 3, 2};
    std::vector<std::vector<int>> exp = {{2, 1, 4, 3, 5}, {3, 2, 1, 4, 5}, {2, 1}};

    auto verify = [&](ListNode* h, std::size_t t) {
        for (int want : exp[t]) {
            if (h == nullptr || h->val != want) return false;
            h = h->next;
        }
        return h == nullptr;
    };

    bool ok1 = true, ok2 = true;

    // ---- Approach 1 ----
    struct rusage r1a, r1b;
    getrusage(RUSAGE_SELF, &r1a);
    auto s1 = std::chrono::high_resolution_clock::now();
    for (std::size_t t = 0; t < vals.size(); ++t) {
        ListNode* head = buildList(vals[t]);
        ListNode* out = Solution_Iterative().reverseKGroup(head, k[t]);
        ok1 = verify(out, t) && ok1;
        deleteList(out);
    }
    auto e1 = std::chrono::high_resolution_clock::now();
    getrusage(RUSAGE_SELF, &r1b);

    // ---- Approach 2 ----
    struct rusage r2a, r2b;
    getrusage(RUSAGE_SELF, &r2a);
    auto s2 = std::chrono::high_resolution_clock::now();
    for (std::size_t t = 0; t < vals.size(); ++t) {
        ListNode* head = buildList(vals[t]);
        ListNode* out = Solution_Recursive().reverseKGroup(head, k[t]);
        ok2 = verify(out, t) && ok2;
        deleteList(out);
    }
    auto e2 = std::chrono::high_resolution_clock::now();
    getrusage(RUSAGE_SELF, &r2b);

    double us1 = std::chrono::duration_cast<std::chrono::nanoseconds>(e1 - s1).count() / 1000.0;
    double us2 = std::chrono::duration_cast<std::chrono::nanoseconds>(e2 - s2).count() / 1000.0;
    long kb1 = (r1b.ru_maxrss - r1a.ru_maxrss) / 1024;
    long kb2 = (r2b.ru_maxrss - r2a.ru_maxrss) / 1024;

    std::cout << "0025 Reverse Nodes in k-Group\n";
    std::cout << "Approach 1 (iterative) : time = " << us1 << " µs  peak-ram-delta = " << kb1 << " KB\n";
    std::cout << "Approach 2 (recursive) : time = " << us2 << " µs  peak-ram-delta = " << kb2 << " KB\n";
    std::cout << (ok1 && ok2 ? "PASS : both approaches regrouped all 3 tests correctly."
                             : "FAIL : at least one test produced a wrong list.")
              << "\n";
    return 0;
}
